### ISSUE TITLE
Exclude admins from username_pattern and make them allowed by default

### ISSUE THREAD

**Issue by GeorgianaElena** [jupyterhub/jupyterhub#3525]
### Proposed change

- Have [`admin_users`](https://github.com/jupyterhub/jupyterhub/blob/main/jupyterhub/auth.py#L87-L102) be part of the [`allowed_users`](https://github.com/jupyterhub/jupyterhub/blob/main/jupyterhub/auth.py#L109-L122) list by default.
- Exclude `admin_users` from [`username_pattern`](https://github.com/jupyterhub/jupyterhub/blob/main/jupyterhub/auth.py#L207-L215)

### Alternative options

Do nothing.

However, consider these scenarios in the current setup:
- only the admin users are allowed to login into the hub
For this to happen the admin usernames need to be both added to `admin_users` list and `allowed_users` list.

- everyone with a `*@uni.edu` account should be allowed into the hub + admins (that may or may not have such an account)
For this to happen, `username_pattern` could be `*@uni.edu`, + all the admins that don't respect this username pattern.

Definitely possible, but they require the extra step of adding the admin usernames to these lists. Not sure if it's just me, but somehow I expected the admins to have these privileges by default.

### Who would use this feature?

Anyone in one of the scenarios above.

### (Optional): Suggest a solution

I would be happy to open a PR for this, but I want to make sure I'm not proposing an anti-pattern :grimacing: 

P.S. I'm unsure of one thing though in terms of implementation: how to keep the first scenario still possible as well as no users apart from admins in `allowed_users` list means anyone is allowed login without adding a new config option.

**Comment by GeorgianaElena** [comment]
Maybe a relevant note for this issue too:

This PR https://github.com/jupyterhub/jupyterhub/pull/3748 and especially [this comment](https://github.com/jupyterhub/jupyterhub/pull/3748#issuecomment-1009526039) from @mathematicalmichael might provide more info about the confusing nature of our current `allowed_users` list.

**Comment by choldgraf** [comment]
I agree - I think that this is a nice way to resolve the problem that @mathematicalmichael's comment describes. I think it is rare that somebody will set up a hub, *not* set any admin users, and also *not* set any allowed_users.

So by making `admin_users` also append itself to `allowed_users`, we'd protect against the case that somebody accidentally provides access to the whole world by only setting `admin_users`.

**Comment by rcthomas** [comment]
We set `admin_users` but don't set any `allowed_users` unless we want to restrict access.  Usually this is something like a test hub with everything else the same as the production hub, but just with `allowed_users` set to be the same list as `admin_users` plus maybe some friendly user testers.  Will that same kind of behavior still be possible?  Right now I think if `admin_users` was simply dropped into `allowed_users` all the time, it would block everyone else.

**Comment by choldgraf** [comment]
Could we add a flag like `authorize_all_users: bool` that would have the same behavior as what you describe, and that would default to `false`?

**Comment by rcthomas** [comment]
I guess that would take some kind of precedence over an empty `allowed_users` list

**Comment by sgibson91** [comment]
I also just want to flag that [these docs](https://jupyterhub.readthedocs.io/en/stable/getting-started/authenticators-users-basics.html#configure-admins-admin-users) say that `c.Authenticator.admin_users` are automatically added to `c.Authenticator.allowed_users` - so there has been some confusion somewhere along the lines!

**Comment by manics** [comment]
@GeorgianaElena What did you decide we should do here?

**Comment by GeorgianaElena** [comment]
Thanks for the ping @manics! I believe next thing here would be to decide if excluding admin usernames from [matching the `username_pattern`](https://github.com/jupyterhub/jupyterhub/blob/main/jupyterhub/auth.py#L248-L261) is a sensible choice or not.

- On one side, I believe it is ok for admins to be excluded from matching this pattern. This is because sometimes hub admins are not members of the institutions most users are part of (and I *think* this is the most common usage of `username_pattern`).
- On the other side, I'm thinking I'm missing some usage cases and this extra privilege for admins is too much?

**Comment by choldgraf** [comment]
When you say "exclude from the `username_pattern` matching" what you mean is "will be allowed access to the hub, even if they do not match `username_pattern`", is that right? If so, then I think this is a reasonable choice - any admin user should also have access to the hub IMO.

**Comment by minrk** [comment]
I'd like to avoid special-casing treatment of admin users, especially as I think we should be trending towards deployments not having any admin users, and instead using role assignments. Anything that only applies to admins is an impediment to that. So anything that applies to 'users' should also apply to admins, who are just users with a specific `admin` role assignment.

**Comment by GeorgianaElena** [comment]
I didn't think  about the roles aspect , but this ⬆️ makes sense to me @minrk! Thanks for explaining it!
I believe I should close this issue, then.
